diff -Naur celeron55-minetest-7244f09-old/src/porting.cpp celeron55-minetest-7244f09/src/porting.cpp
--- celeron55-minetest-7244f09-old/src/porting.cpp	2011-09-22 21:13:09.279654700 +0200
+++ celeron55-minetest-7244f09/src/porting.cpp	2011-09-22 21:15:51.648411215 +0200
@@ -209,13 +209,7 @@
 	
 	pathRemoveFile(buf, '/');
 
-	path_data = std::string(buf) + "/../share/" + PROJECT_NAME;
-	//path_data = std::string(INSTALL_PREFIX) + "/share/" + PROJECT_NAME;
-	if (!fs::PathExists(path_data)) {
-		dstream<<"WARNING: data path " << path_data << " not found!";
-		path_data = std::string(buf) + "/../data";
-		dstream<<" Trying " << path_data << std::endl;
-	}
+	path_data = std::string(buf) + "/../../share/games/" + PROJECT_NAME;
 	
 	path_userdata = std::string(getenv("HOME")) + "/." + PROJECT_NAME;
 
